src/storage/src/manifest/region.rs

```diff
@@ -13,10 +13,146 @@
 // limitations under the License.
 
 //! Region manifest impl
+use std::any::Any;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use common_telemetry::info;
+use object_store::ObjectStore;
+use store_api::manifest::action::ProtocolAction;
+use store_api::manifest::{
+    Manifest, ManifestLogStorage, ManifestVersion, MetaActionIterator, MIN_VERSION,
+};
+
+use crate::error::Result;
 use crate::manifest::action::*;
+use crate::manifest::checkpoint::Checkpointer;
 use crate::manifest::ManifestImpl;
 
-pub type RegionManifest = ManifestImpl<RegionMetaActionList>;
+pub type RegionManifest = ManifestImpl<RegionSnapshot, RegionMetaActionList>;
+
+#[derive(Debug)]
+pub struct RegionManifestCheckpointer {
+    flushed_manifest_version: AtomicU64,
+}
+
+impl RegionManifestCheckpointer {
+    pub(crate) fn set_flushed_manifest_version(&self, manifest_version: ManifestVersion) {
+        self.flushed_manifest_version
+            .store(manifest_version, Ordering::Relaxed);
+    }
+}
+
+#[async_trait]
+impl Checkpointer for RegionManifestCheckpointer {
+    type Snapshot = RegionSnapshot;
+    type MetaAction = RegionMetaActionList;
+
+    async fn do_checkpoint(
+        &self,
+        manifest: &ManifestImpl<RegionSnapshot, RegionMetaActionList>,
+    ) -> Result<Option<RegionSnapshot>> {
+        let last_snapshot = manifest.last_snapshot().await?;
+
+        let current_version = manifest.last_version();
+        let (start_version, mut protocol, mut manifest_builder) =
+            if let Some(snapshot) = last_snapshot {
+                (
+                    snapshot.last_version,
+                    snapshot.protocol,
+                    RegionManifestDataBuilder::with_snapshot(snapshot.snapshot),
+                )
+            } else {
+                (
+                    MIN_VERSION,
+                    ProtocolAction::default(),
+                    RegionManifestDataBuilder::default(),
+                )
+            };
+
+        // Checkpoint can't exceed over flushed manifest version.
+        // We have to keep the region metadata which are not flushed for replaying WAL.
+        let end_version =
+            (current_version + 1).min(self.flushed_manifest_version.load(Ordering::Relaxed) + 1);
+        if start_version + 1 >= end_version {
+            return Ok(None);
+        }
+
+        let mut iter = manifest.scan(start_version, end_version).await?;
```

Comment:
end + 1 for `[start, end)` (and corresponding to the previous change). These two suggestions are non-blocking and feel free to ignore them.

```suggestion
        let mut iter = manifest.scan(start_version, end_version + 1).await?;
```

Reply:
I prefer to ensure the `end_version` is exclusive.